packages/codemirror-blocks/src/components/DropTarget.tsx

```diff
@@ -145,30 +156,46 @@ field declared. The node was:`,
   const [mouseOver, setMouseOver] = useState(false);
 
   const dispatch: AppDispatch = useDispatch();
-  const ast = useSelector(selectors.getAST);
-  const isEditable = useSelector(selectors.getEditable)[id] ?? false;
+
+  const createTarget = useCallback(
+    (ast: AST) =>
+      new actions.InsertTarget(
+        node,
+        props.field,
+        getLocation({
+          id,
+          ast,
+          context: {
+            field: props.field,
+            node,
+          },
+        })
+      ),
+    [id, node, props.field]
+  );
+
+  // create and use a new insertTarget selector for evey
+  // DropTarget instance to avoid re-rerendering every
+  // drop target when any other drop target's editable
+  // state changes. The insertTarget will be null if
+  // the node is not editable.
+  const insertTarget = useSelector(
```

Comment:
🤯 This is the part I'm trying to wrap my brain around. Here's what I _think_ this changes is all about:

So in the previous implementation, we'd create a new insertionTarget on every render, using state from the redux store obtained via `dispatch`. And since it was a new object each time, React would re-render it each time.

The use of `dispatch` can be replaced with a call to `useSelector`, which is what's happening deeper inside the new function. That's a style change, bringing this code into conformance with the rest of the codebase but not providing an optimization -- because the selector itself would also be new on each render. But you've optimized this by creating a _memoized_ selector.

Do I have this right?

Reply:
Close. `createTarget` requires the ast which comes from the redux store. However, we weren't using `dispatch` to get that state before. We were however using `useSelector` to subscribe to all changes to the ast, meaning this component would re-render on any change to the AST. But actually, the component only needs to rerender on a change to the list of editable nodes.

The additional `useMemo` part is because the memoization that `createSelector` already does only has a cache size of one, so it doesn't work so well for a selector that takes a parameter (in this case the `id`). The react-redux api for this is more complicated than I'd like it to be, but this is apparently what they recommend for these situations according to these docs: https://react-redux.js.org/api/hooks#using-memoizing-selectors

Reply:
> But actually, the component only needs to rerender on a change to the list of editable nodes.
Does this mean that _every_ dropTarget re-renders whenever the list of editable nodes changes?

Clearly that's a huge improvement, but this is making think about possible ways to optimize further. For example, a node cannot become editable unless it's focused....